packages/wrangler/src/deployment-bundle/esbuild-plugins/nodejs-compat.ts

```diff
@@ -1,13 +1,83 @@
+import { readFile } from "fs/promises";
+import chalk from "chalk";
+import { logger } from "../../logger";
 import type { Plugin } from "esbuild";
 
+// Infinite loop detection
+const seen = new Set();
+
+// Prevent multiple warnings per package
+const warnedPackaged = new Map();
+
 /**
  * An esbuild plugin that will mark any `node:...` imports as external.
  */
-export const nodejsCompatPlugin: Plugin = {
+export const nodejsCompatPlugin: (silenceWarnings: boolean) => Plugin = (
+	silenceWarnings
+) => ({
 	name: "nodejs_compat imports plugin",
 	setup(pluginBuild) {
-		pluginBuild.onResolve({ filter: /node:.*/ }, () => {
-			return { external: true };
+		pluginBuild.onResolve(
+			{ filter: /node:.*/ },
+			async ({ path, kind, resolveDir, ...opts }) => {
+				if (seen.has(`${path}:${kind}:${resolveDir}:${opts.importer}`)) {
+					return;
+				}
+
+				seen.add(`${path}:${kind}:${resolveDir}:${opts.importer}`);
+				// Try to resolve this import as a normal package
+				const result = await pluginBuild.resolve(path, {
+					kind,
+					resolveDir,
+					importer: opts.importer,
+				});
+
+				if (result.errors.length > 0) {
+					// esbuild couldn't resolve the package
+					// We should warn the user, but not fail the build
+
+					// First, check whether the importer has specifically silenced warnings for this import
+					const importer = await readFile(opts.importer, "utf8");
+
+					const allowlistedModules =
+						/\/\/ wrangler-allowlist (".*?")\w*\n/.exec(importer);
+					if (allowlistedModules?.[1]) {
+						// Remove quotes
+						const modules = allowlistedModules[1]
+							.split(" ")
+							.map((m) => m.slice(1, -1));
+
+						// Disable warnings for this node:* module in this file
+						if (modules.includes(path)) {
+							return { external: true };
+						}
+					}
+
+					if (!warnedPackaged.has(path)) {
+						warnedPackaged.set(path, [opts.importer]);
+					} else {
+						warnedPackaged.set(path, [
+							...warnedPackaged.get(path),
+							opts.importer,
+						]);
+					}
+					return { external: true };
+				}
+				// This is a normal package—don't treat it specially
+				return result;
+			}
+		);
+		// Wait until the build finishes to log warnings, so that all files which import a package
+		// can be collated
+		pluginBuild.onEnd(() => {
+			if (!silenceWarnings)
+				warnedPackaged.forEach((importers: string[], path: string) => {
```

Comment:
This warning won't be logged if `node_compat = true` is set, since the polyfill plugin runs first. If you have the polyfills enabled, you'd see:

```
▲ [WARNING] Enabling Node.js compatibility mode for built-ins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details.
```

(the same as right now)

And if you have the Node.JS compat flag enabled you'd see no warnings